Client/Frontend/Browser/TabManagerNavDelegate.swift

```diff
@@ -78,47 +78,56 @@ class TabManagerNavDelegate: NSObject, WKNavigationDelegate {
-        })
+      // Needed to resolve ambiguous delegate signatures: https://github.com/apple/swift/issues/45652#issuecomment-1149235081
+      typealias WKNavigationActionSignature = (WKNavigationDelegate) -> ((WKWebView, WKNavigationAction, WKWebpagePreferences, @escaping (WKNavigationActionPolicy, WKWebpagePreferences) -> Void) -> Void)?
+      
+      // Needed to detect if async implementations exist as we cannot detect them directly
+      if delegate.responds(to: #selector(WKNavigationDelegate.webView(_:decidePolicyFor:preferences:decisionHandler:) as WKNavigationActionSignature)) {
+        // Do NOT change to `delegate.webView?(....)` the optional operator makes async-await calls crash the compiler atm!
+        // It must be force-unwrapped at the time of writing `January 10th, 2023`.
+        let (policy, preferences) = await delegate.webView!(webView, decidePolicyFor: navigationAction, preferences: preferences)
+        if policy == .cancel {
+          res = policy
+        }
+        
+        pref = preferences
+      }
     }
-
-    decisionHandler(res, pref)
+    
+    return (res, pref)
   }
-
-  func webView(
-    _ webView: WKWebView,
-    decidePolicyFor navigationResponse: WKNavigationResponse,
-    decisionHandler: @escaping (WKNavigationResponsePolicy) -> Void
-  ) {
+  
+  func webView(_ webView: WKWebView, decidePolicyFor navigationResponse: WKNavigationResponse) async -> WKNavigationResponsePolicy {
     var res = WKNavigationResponsePolicy.allow
     for delegate in delegates {
-      delegate.webView?(
-        webView, decidePolicyFor: navigationResponse,
-        decisionHandler: { policy in
-          if policy == .cancel {
-            res = policy
-          }
-        })
+      // Needed to resolve ambiguous delegate signatures: https://github.com/apple/swift/issues/45652#issuecomment-1149235081
+      typealias WKNavigationResponseSignature = (WKNavigationDelegate) -> ((WKWebView, WKNavigationResponse, @escaping (WKNavigationResponsePolicy) -> Void) -> Void)?
+      
+      // Needed to detect if async implementations exist as we cannot detect them directly
+      if delegate.responds(to: #selector(WKNavigationDelegate.webView(_:decidePolicyFor:decisionHandler:) as WKNavigationResponseSignature)) {
+        // Do NOT change to `delegate.webView?(....)` the optional operator makes async-await calls crash the compiler atm!
+        // It must be force-unwrapped at the time of writing `January 10th, 2023`.
+        let policy = await delegate.webView!(webView, decidePolicyFor: navigationResponse)
+        if policy == .cancel {
+          res = policy
+        }
+      }
     }
 
     if res == .allow {
-      let tab = tabManager?[webView]
-      tab?.mimeType = navigationResponse.response.mimeType
+      // TabManager.subscript.getter requires MAIN-THREAD!
+      await Task { @MainActor in
+        let tab = tabManager?[webView]
+        tab?.mimeType = navigationResponse.response.mimeType
+      }.value
```

Comment:
`@MainActor` can be added to the function definition instead of doing this.

Also for future usage: When you don't have any async code executing and you want it on main you can do `await MainActor.run { ... }` instead of `await Task { @MainActor in ... }`

Reply:
Also if adding `@MainActor` here you may want to also add it to BVC's async delegate methods too if that was the original place it was always run